osc-server/src/main/java/org/osc/core/broker/service/DeleteDeploymentSpecService.java

```diff
@@ -113,18 +118,38 @@ private void validateAndLoad(EntityManager em, BaseDeleteRequest request) throws
             + " is not found.");
         }
 
-        if (DeploymentSpecEntityMgr.isProtectingWorkload(this.ds) ||
-                VirtualSystemEntityMgr.isProtectingWorkload(vs)) {
+        if (DeploymentSpecEntityMgr.isProtectingWorkload(this.ds)) {
             throw new VmidcBrokerValidationException(
                     String.format("The deployment spec with name '%s' is currently protecting a workload",
                             this.ds.getName()));
         }
 
+        //check if virtual system of this DS is attached to a SFC and only allow deletion of DS if no SG is binding
+        // to this virtual systems SFC.
+        checkSfcAndBinding(em, vs);
+
         if (!this.ds.getMarkedForDeletion() && request.isForceDelete()) {
             throw new VmidcBrokerValidationException(
                     "Deployment Spec '"
                             + this.ds.getName()
                             + "' is not marked for deletion and force delete operation is applicable only for entries marked for deletion.");
         }
     }
+
+    private void checkSfcAndBinding(EntityManager em, VirtualSystem vs) throws Exception {
+        if (!vs.getServiceFunctionChains().isEmpty()) {
+            //Look if these SFCs are binded to any of the SG of same tenant as of DS
+            for (ServiceFunctionChain sfc : vs.getServiceFunctionChains()) {
+                List<SecurityGroup> sgList = SecurityGroupEntityMgr.listSecurityGroupsBySfcIdAndProjectId(em,
+                        sfc.getId(), this.ds.getProjectId());
+                SecurityGroup sgActive = sgList.stream().filter(sg -> !sg.getMarkedForDeletion()).findFirst()
```

Comment:
you can give all the sg names comma seperated.

stream()
                .filter(sg -> !sg.getMarkedForDeletion()).map(sg -> sg.getName()).collect(Collectors.joining(", "))